### PR TITLE
Download media from given web contents

### PR DESCRIPTION
**Pull request by sangwoo108**
In real situation, we'll download medias from
web contents in tab strip model. In order to demonstrate that, add API and test UI for brave://playlist.

Resolves https://github.com/brave/brave-browser/issues/24405

DEMO:

https://user-images.githubusercontent.com/5474642/182630842-86b93ba2-65ad-4f18-b8ed-1919ad0dac07.mov

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/playlist/playlist_tab_helper.cc [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        delete window.WebKitMediaSource;
+      }
+    })();
+    )-";
+
+  // In order to hide js API from main world, use testing
+  // api temporarily.
+  const auto execute_script_callback = [](std::u16string script,
+                                          content::RenderFrameHost* frame) {
+    frame->ExecuteJavaScriptForTests(kScriptToHideMediaSourceAPI,
```
Comment (fallaciousreasoning) [inline components/playlist/playlist_tab_helper.cc:46]: I guess this is fine because it's behind a flag for now (and we were doing this before anyway)

**Review by fallaciousreasoning — COMMENTED** [review]
Mostly looks pretty good - however, I was trying to run it locally but I couldn't get it to work

File: components/playlist/playlist_service.cc [hunk 2]
```diff
@@ -78,12 +78,28 @@ void PlaylistService::Shutdown() {
   task_runner_.reset();
 }
 
+void PlaylistService::RequestDownloadMediaFilesFromContents(
+    const std::string& playlist_id,
+    content::WebContents* contents) {
+  VLOG(2) << __func__
+          << " download medias from WebContents to playlist: " << playlist_id;
```
[the remark this answers is not in the document]
Comment (sangwoo108) [reply components/playlist/playlist_service.cc:85]: Thanks! Done.

File: components/playlist/playlist_tab_helper.cc [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    : content::WebContentsObserver(contents),
+      content::WebContentsUserData<PlaylistTabHelper>(*contents) {}
+
+PlaylistTabHelper::~PlaylistTabHelper() = default;
+
+// static
+void PlaylistTabHelper::HideMediaSrcAPIFromContents(
+    content::WebContents* contents) {
+  if (auto url = contents->GetVisibleURL();
+      url.scheme() == "chrome-extension" || url.host_piece() == "127.0.0.1") {
```
[the remark this answers is not in the document]
Comment (sangwoo108) [reply components/playlist/playlist_tab_helper.cc:20]: Yeah, I'm not a fan of this sort of blocked list. I'm looking for a way to check if it's testing at runtime.

Or maybe it'd be better to run this script only for limited sites known to using blob: src. (typically youtube)

File: components/playlist/playlist_tab_helper.cc (hunk 1 above)
Comment (sangwoo108) [reply components/playlist/playlist_tab_helper.cc:46]: Yeah, This is another thing I'm not sure about. AFAIK, ExecutingJavaScript(), which runs in main world, is allowed on Android, but not Desktop. I guess why they allow it for Android is that it isn't supported to make extension API for browser on Android. And iOS folks seems to be doing similar things with WKWebView. So my question is, should we avoid this since/if we can do this with an extension? I'm not sure. @bsclifton @simonhong @petemill  Do you think there's a sort of security reason for preventing ExecuteJavascript in main world on desktop?

**Review by sangwoo108 — COMMENTED** [review]
> was trying to run it locally but I couldn't get it to work

That's weired :( I don't remember implementing country lock. Let me check it again

File: browser/brave_tab_helpers.cc [hunk 4]
```diff
@@ -66,6 +67,10 @@
 #include "brave/browser/web_discovery/web_discovery_tab_helper.h"
 #endif
 
+#if BUILDFLAG(ENABLE_PLAYLIST)
```
[the remark this answers is not in the document]
Comment (sangwoo108) [reply browser/brave_tab_helpers.cc:70]: I presumed Android folks will need PlaylistTabHelpers.

File: components/playlist/playlist_tab_helper.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (sangwoo108) [reply components/playlist/playlist_tab_helper.cc]: Changed to return when the feature is disabled.

File: browser/brave_tab_helpers.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (fallaciousreasoning) [reply browser/brave_tab_helpers.cc:70]: :+1:

File: browser/ui/webui/playlist_page_handler.cc [hunk 5]
```diff
@@ -72,6 +76,26 @@ void PlaylistPageHandler::AddMediaFilesFromPageToPlaylist(const std::string& id,
...
 
+void PlaylistPageHandler::AddMediaFilesFromOpenTabsToPlaylist(
+    const std::string& playlist_id) {
+  std::vector<content::WebContents*> contentes;
+
+  auto* browser = chrome::FindBrowserWithWebContents(web_contents_);
+  DCHECK(browser);
+  auto* tab_strip_model = browser->tab_strip_model();
+  for (auto i = 0; i < tab_strip_model->count(); i++) {
+    if (auto* contents = tab_strip_model->GetWebContentsAt(i);
```
[the remark this answers is not in the document]
Comment (fallaciousreasoning) [reply browser/ui/webui/playlist_page_handler.cc:87]: Sweet cool! Reducing the scope is pretty cool :)

File: components/playlist/playlist_download_request_manager.cc [hunk 6]
```diff
@@ -240,12 +238,13 @@ void PlaylistDownloadRequestManager::OnGetMedia(base::Value value) {
...
-      FROM_HERE, kWebContentDestroyDelay,
-      base::BindOnce(&PlaylistDownloadRequestManager::DestroyWebContents,
-                     base::Unretained(this)));
+  if (!web_contents_destroy_timer_) {
+    web_contents_destroy_timer_ = std::make_unique<base::RetainingOneShotTimer>(
+        FROM_HERE, kWebContentDestroyDelay,
+        base::BindRepeating(&PlaylistDownloadRequestManager::DestroyWebContents,
+                            base::Unretained(this)));
+  }
+  web_contents_destroy_timer_->Reset();
```
[the remark this answers is not in the document]
Comment (fallaciousreasoning) [reply components/playlist/playlist_download_request_manager.cc:247]: Ah right, my bad :facepalm:

File: components/playlist/playlist_tab_helper.cc (hunk 1 above)
Comment (fallaciousreasoning) [reply components/playlist/playlist_tab_helper.cc:51]: Yeah, that makes sense :smile: I guess we'll get someone with a bit more authority than me to make that call :+1:

**Review by fallaciousreasoning — COMMENTED** [review]
FWIW, this LGTM. However, I still can't get it to run locally (looking at the console output, it's failing to load `cr.js`). I feel like I probably don't have something configured that I should :laughing: 

image

I also think we should get someone a bit heavier duty than me to way in on the `ExecuteJavascript` use.

File: components/playlist/playlist_tab_helper.cc (hunk 1 above)
Comment (sangwoo108) [reply components/playlist/playlist_tab_helper.cc:51]: Yep, I'm going to ask other folks about this! Also, I'm thinking of making something like `ScopedAllowingJavascriptExecution` which allows JS execution for short time.

**Comment by sangwoo108** [comment]
> just tested in Nightly, and playlist does work for me there, so it's probably something about my local

Thanks @fallaciousreasoning ! When I cleared user data dir, it happens to me too! I think there's a problem in component. Let me check out again.

File: browser/ui/webui/playlist_page_handler.cc [hunk 7]
```diff
@@ -72,6 +76,26 @@ void PlaylistPageHandler::AddMediaFilesFromPageToPlaylist(const std::string& id,
...
+void PlaylistPageHandler::AddMediaFilesFromOpenTabsToPlaylist(
+    const std::string& playlist_id) {
+  std::vector<content::WebContents*> contentses;
+
+  auto* browser = chrome::FindBrowserWithWebContents(web_contents_);
+  DCHECK(browser);
+  auto* tab_strip_model = browser->tab_strip_model();
+  for (auto i = 0; i < tab_strip_model->count(); i++) {
+    if (auto* contents = tab_strip_model->GetWebContentsAt(i);
+        contents != web_contents_) {
```
Comment (simonhong) [inline browser/ui/webui/playlist_page_handler.cc:88]: IMO, this checking could be improved. Instead of comparing web contents itself, how about checking whether it loaded supported service (ex, youtube.com) or not?
Crash happened when this api is called when there are no other tabs that loaded youtube.com.

Comment (sangwoo108) [reply browser/ui/webui/playlist_page_handler.cc:88]: > IMO, this checking could be improved. Instead of comparing web contents itself, how about checking whether it loaded supported service (ex, youtube.com) or not?

Oops, this isn't a thing to filter service. This is temporal code to avoid targeting `brave://playlist` itself. The crash seems to be from CHECK(). That should be handled in other way.

Comment (sangwoo108) [reply browser/ui/webui/playlist_page_handler.cc:88]: Thank you for letting me know the crash. It should be gone now!

**Comment by fallaciousreasoning** [comment]
Interesting, the folder `~/.config/BraveSoftware/Brave-Browser-Development/jccpmjhflblpphnhgemhlllckflnipjn` exists but it's empty on my machine. 

I tested with a new profile and the same thing happened (`/tmp/bar/jccpmjhflblpphnhgemhlllckflnipjn` exists but is empty).

**Comment by sangwoo108** [comment]
> Interesting, the folder `~/.config/BraveSoftware/Brave-Browser-Development/jccpmjhflblpphnhgemhlllckflnipjn` exists but it's empty on my machine.
> 
> I tested with a new profile and the same thing happened (`/tmp/bar/jccpmjhflblpphnhgemhlllckflnipjn` exists but is empty).

I should set up linux machine. @simonhong Have you ever heard of component problem on Linux?

**Comment by fallaciousreasoning** [comment]
Also though, it happens on master, so I don't think it should block this PR :smile:

**Comment by simonhong** [comment]
> > Interesting, the folder `~/.config/BraveSoftware/Brave-Browser-Development/jccpmjhflblpphnhgemhlllckflnipjn` exists but it's empty on my machine.
> > I tested with a new profile and the same thing happened (`/tmp/bar/jccpmjhflblpphnhgemhlllckflnipjn` exists but is empty).
> 
> I should set up linux machine. @simonhong Have you ever heard of component problem on Linux?

Let me check this PR on my linux.

**Comment by simonhong** [comment]
I confirmed that component is properly downloaded.
Interesting thing is only `click here to download` works. `Download media from open tabs` doesn't work on linux.
Both works well on macOS.

File: components/playlist/playlist_download_request_manager.cc [hunk 8]
```diff
@@ -115,10 +115,28 @@ void PlaylistDownloadRequestManager::RunMediaDetector(Request request) {
...
+      FetchPendingRequest();
+      return;
+    }
+
+    DCHECK(weak_contents->GetMainFrame());
+    weak_contents->GetMainFrame()->ExecuteJavaScriptInIsolatedWorld(
+        base::UTF8ToUTF16(media_detector_script_),
+        base::BindOnce(&PlaylistDownloadRequestManager::OnGetMedia,
+                       weak_factory_.GetWeakPtr()),
+        g_playlist_javascript_world_id);
```
Comment (simonhong) [inline components/playlist/playlist_download_request_manager.cc:138]: How about making this as a separated method like `GetMedia(WebContents* web_contents);` and share with `DidFinishLoad()`?

File: components/playlist/playlist_download_request_manager.cc [hunk 9]
```diff
@@ -175,6 +173,10 @@ void PlaylistDownloadRequestManager::OnGetMedia(base::Value value) {
     LOG(ERROR) << value;
```
Comment (diracdeltas) [inline components/playlist/playlist_download_request_manager.cc:173]: not technically part of this PR but shouldn't this be VLOG?

also if this is happening in the main process, seems `value` should be passed through the JSON sanitizer before calling  `base::JSONWriter::Write` to comply with https://chromium.googlesource.com/chromium/src/+/master/docs/security/rule-of-2.md

File: components/playlist/playlist_download_request_manager.cc [hunk 10]
```diff
@@ -115,10 +115,28 @@ void PlaylistDownloadRequestManager::RunMediaDetector(Request request) {
...
-  GURL url(request.url);
-  DCHECK(url.is_valid());
-  web_contents_->GetController().LoadURLWithParams(
-      content::NavigationController::LoadURLParams(url));
+  if (absl::holds_alternative<std::string>(request.url_or_contents)) {
+    CreateWebContents();
+    GURL url(absl::get<std::string>(request.url_or_contents));
+    DCHECK(url.is_valid());
+    DCHECK(web_contents_);
+    web_contents_->GetController().LoadURLWithParams(
```
Comment (diracdeltas) [inline components/playlist/playlist_download_request_manager.cc:123]: does this get loaded in incognito/tor contexts for playlist requests associated with incognito/tor windows?

File: components/playlist/playlist_tab_helper.cc (hunk 1 above)
Comment (diracdeltas) [reply components/playlist/playlist_tab_helper.cc:50]: we do execute js in the main world for instance in wallet, but there are some security concerns like the main page being able to do prototype pollution on the script. cc @bridiver in case he knows a better way here

Comment (sangwoo108) [reply components/playlist/playlist_tab_helper.cc:50]: Thanks @diracdeltas . Then can I merge this PR or should I wait for overall review?

File: browser/brave_tab_helpers.cc [hunk 11]
```diff
@@ -124,6 +129,10 @@ void AttachTabHelpers(content::WebContents* web_contents) {
   }
 
   brave_wallet::BraveWalletTabHelper::CreateForWebContents(web_contents);
+
+#if BUILDFLAG(ENABLE_PLAYLIST)
+  playlist::PlaylistTabHelper::CreateForWebContents(web_contents);
```
[the remark this answers is not in the document]
Comment (sangwoo108) [reply browser/brave_tab_helpers.cc]: Yep. Done.

**Review by simonhong — APPROVED** [review]

File: components/playlist/playlist_download_request_manager.cc (hunk 10 above)
Comment (diracdeltas) [reply components/playlist/playlist_download_request_manager.cc:123]: @sangwoo108 i strongly feel we should make sure there is no way to enable playlist UI in incognito/tor windows. if i'm visiting a site in Tor, i probably never want anything associated with the site to be loaded without the Tor proxy. if the UI is enabled in Tor windows, all the requests for content associated with the Tor window should go through Tor.

File: components/playlist/playlist_tab_helper.cc (hunk 1 above)
Comment (diracdeltas) [reply components/playlist/playlist_tab_helper.cc:50]: let me ping brian to take a look first

File: browser/brave_content_browser_client.cc [hunk 12]
```diff
@@ -218,9 +218,9 @@ using extensions::ChromeContentBrowserClientExtensionsPart;
 #include "brave/browser/brave_ads/brave_ads_host.h"
 #endif  // BUILDFLAG(IS_ANDROID)
 
-#if BUILDFLAG(ENABLE_PLAYLIST)
+#if BUILDFLAG(ENABLE_PLAYLIST) && !BUILDFLAG(IS_ANDROID)
```
Comment (bridiver) [inline browser/brave_content_browser_client.cc:221]: can we add a new buildflag enable_playlist_webui or something?

File: browser/brave_tab_helpers.cc [hunk 13]
```diff
@@ -124,6 +130,13 @@ void AttachTabHelpers(content::WebContents* web_contents) {
   }
 
   brave_wallet::BraveWalletTabHelper::CreateForWebContents(web_contents);
+
+#if BUILDFLAG(ENABLE_PLAYLIST)
+  if (playlist::PlaylistServiceFactory::IsPlaylistEnabled(
```
Comment (bridiver) [inline browser/brave_tab_helpers.cc:135]: it's better to have `MaybeCreateForWebContents` and put this inside

File: browser/ui/webui/playlist_page_handler.cc [hunk 14]
```diff
@@ -72,6 +76,20 @@ void PlaylistPageHandler::AddMediaFilesFromPageToPlaylist(const std::string& id,
                                                                   url.spec());
 }
 
+void PlaylistPageHandler::AddMediaFilesFromOpenTabsToPlaylist(
```
Comment (bridiver) [inline browser/ui/webui/playlist_page_handler.cc:79]: I'm not sure I understand what's happening here, we're adding all media files from open tabs except this tab?

Comment (bridiver) [reply browser/ui/webui/playlist_page_handler.cc:79]: oh I see, web_contents_ is the webui webcontents. This is not correct then because the webui webcontents is not associated with a browser

File: components/playlist/playlist_tab_helper.cc (hunk 1 above)
Comment (bridiver) [reply components/playlist/playlist_tab_helper.cc:50]: this definitely needs to be changed and the same with the call in PlaylistDownloadRequestManager

Comment (bridiver) [reply components/playlist/playlist_tab_helper.cc:50]: also it is not going to work reliably anyway. You are making an async call to run JS so you have a race condition. @antonok-edm can we make this work through cf somehow?

Comment (bridiver) [reply components/playlist/playlist_tab_helper.cc:50]: if we can't enable it through cf, this needs to be a RenderFrameObserver, not a WebContentsObserver

File: browser/ui/webui/playlist_page_handler.cc (hunk 14 above)
Comment (sangwoo108) [reply browser/ui/webui/playlist_page_handler.cc:79]: > webui webcontents is not associated with a browser

Sorry, I'm not sure what you mean here. Currently, `brave://playlist` is loaded on a tab and I thought we should filter it from contents of tab strip model.

File: components/playlist/playlist_tab_helper.cc (hunk 1 above)
Comment (sangwoo108) [reply components/playlist/playlist_tab_helper.cc:50]: May I ask you what `cf` is?

File: components/playlist/playlist_download_request_manager.cc (hunk 10 above)
Comment (sangwoo108) [reply components/playlist/playlist_download_request_manager.cc:123]: Thanks! I'll make sure that.

File: browser/brave_tab_helpers.cc (hunk 13 above)
Comment (sangwoo108) [reply browser/brave_tab_helpers.cc:135]: Humm, that might bring a layering violation.

```
// static
PlaylistTabHelper::MaybeCreateForWebContents() { << PlaylistTabHelper is in component layer
  if (PlaylistServiceFactory::IsPlaylistEnabled()) { << PlaylistServiceFactory is in browser layer.
  }
}
```

Or, do you think it's still better to have the function in anonymous namespace in this file?

File: browser/brave_content_browser_client.cc (hunk 12 above)
Comment (sangwoo108) [reply browser/brave_content_browser_client.cc]: Sounds good! Done.

File: components/playlist/playlist_tab_helper.cc (hunk 1 above)
Comment (sangwoo108) [reply components/playlist/playlist_tab_helper.cc:50]: It seems we can turn off MediaSourceAPI with RuntimeEnabledFeature. Then we can remove this kind of unreliable js call. Let me try again.

File: browser/brave_tab_helpers.cc (hunk 13 above)
Comment (goodov) [reply browser/brave_tab_helpers.cc:135]: you can either put tab_helper into `src/brave/playlist` or make `PlaylistServiceFactory::IsPlaylistEnabled` as a `component/playlist/...`-located function.

File: components/playlist/playlist_tab_helper.cc [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+#endif
+  };
+
+  contents->GetMainFrame()->ForEachRenderFrameHost(base::BindRepeating(
+      execute_script_callback, kScriptToHideMediaSourceAPI));
+}
+
+void PlaylistTabHelper::DidFinishNavigation(
+    content::NavigationHandle* navigation_handle) {
+  HideMediaSrcAPIFromContents(web_contents());
```
Comment (goodov) [inline components/playlist/playlist_tab_helper.cc:61]: this is now called for all navigations, including non-main frame ones and same document ones.
this is called for all tabs even if playlist feature wasn't used at all (just enabled).

why?

File: components/playlist/playlist_tab_helper.cc (hunk 1 above)
Comment (goodov) [reply components/playlist/playlist_tab_helper.cc:50]: I assume `cf` is `cosmetic filters`.
we shouldn't have `*ForTests` anywhere in production code. One of other options is to implement a renderer-based class and call it via mojo to run all playlist-related main frame JS there.
I assume the current approach might break websites in some way, because it's not synchronous, i.e. the time between completed navigation and a requested JS execution might be pretty big.

File: browser/brave_tab_helpers.cc (hunk 13 above)
Comment (sangwoo108) [reply browser/brave_tab_helpers.cc:135]: `PlaylistTabHelper` was located in the component layer as it might be used for Android too. And PlaylistServiceFactory seems to be located in browser layer cause @simonhong didn't want to make the component depend on Profile. I think moving PlaylistServiceFactory into the component worth trying.

File: components/playlist/playlist_tab_helper.cc (hunk 15 above)
Comment (sangwoo108) [reply components/playlist/playlist_tab_helper.cc:61]: If we don't hide MediaSource API ahead of media detection script, we will get `blob:` source URL, which we can't download from.

**Comment by goodov** [comment]
@sangwoo108 please don't force push when a code-review is already in progress. I have no idea what has changed since last time :)
IMO it's okay to force push when a PR is just created and you're fixing obvious build issues, but when there's a lot of discussion on critical parts, let's use new commits to add new things, otherwise it's very hard to keep track of what's going on.

File: components/playlist/playlist_tab_helper.cc (hunk 15 above)
Comment (goodov) [reply components/playlist/playlist_tab_helper.cc]: my point is that it will happen in all pages, even if user doesn't care about playlists in any way. Going this path we just can disable these endpoints in blink, but I don't think this is what we want.

File: browser/brave_tab_helpers.cc (hunk 13 above)
Comment (goodov) [reply browser/brave_tab_helpers.cc]: factories should always be located in brave/browser. Profile is `//chrome/browser` concept, not `//components` one.

File: components/playlist/playlist_tab_helper.cc (hunk 15 above)
Comment (sangwoo108) [reply components/playlist/playlist_tab_helper.cc]: > Going this path we just can disable these endpoints in blink, but I don't think this is what we want.

Aha. I see where you're coming from now. We're going to support a global preference or content setting for this behavior. But for now we don't have any specific plan.

**Comment by sangwoo108** [comment]
> @sangwoo108 please don't force push when a code-review is already in progress. I have no idea what has changed since last time :) IMO it's okay to force push when a PR is just created and you're fixing obvious build issues, but when there's a lot of discussion on critical parts, let's use new commits to add new things, otherwise it's very hard to keep track of what's going on.

Oops, sorry. I didn't notice that.

File: browser/brave_tab_helpers.cc (hunk 13 above)
Comment (sangwoo108) [reply browser/brave_tab_helpers.cc]: > Profile is //chrome/browser concept, not //components one

Oh, yeah. I was mistaken about it. I tought there's a something like //component/profile

File: components/playlist/playlist_tab_helper.cc (hunk 1 above)
Comment (sangwoo108) [reply components/playlist/playlist_tab_helper.cc]: Fortunately, RuntimeEnabledFeature works as we needed. I've removed JS injection part and MediaSource api will be disabled from blink side. But disabling should be improved based on preference or content setting in the future.

File: chromium_src/content/child/runtime_features.cc [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+void SetRuntimeFeaturesDefaultsAndUpdateFromArgs(
+    const base::CommandLine& command_line) {
+  SetRuntimeFeaturesDefaultsAndUpdateFromArgs_ChromiumImpl(command_line);
+
+#if BUILDFLAG(ENABLE_PLAYLIST)
+  SetRuntimeFeatureFromChromiumFeature(
+      playlist::features::kPlaylist, kDefault, [](bool enabled) {
+        WebRuntimeFeatures::EnableFeatureFromString("MediaSourceStable",
+                                                    !enabled);
+      });
```
Comment (sangwoo108) [inline chromium_src/content/child/runtime_features.cc:28]: @bridiver @goodov This is the new way to MediaSource API without JS injection. What do you think?

File: patches/content-child-BUILD.gn.patch [hunk 17]
```diff
@@ -0,0 +1,12 @@
+diff --git a/content/child/BUILD.gn b/content/child/BUILD.gn
+index 39264fc8a19083aeee6c65e1b4f3dd4d4da72aea..4b2bad00f3a4365eb7967eba2a73a228dedeced7 100644
+--- a/content/child/BUILD.gn
++++ b/content/child/BUILD.gn
+@@ -66,6 +66,7 @@ target(link_target_type, "child") {
+   deps = [
+     "//base",
+     "//build/config/compiler:compiler_buildflags",
++    "//brave/components/playlist:playlist_features",
```
Comment (simonhong) [inline patches/content-child-BUILD.gn.patch:9]: How about doing this with `sources.gni` like below?
With that, we don't need to touch this patch anymore when we want to add more deps or sources to this target in the future.
https://github.com/brave/brave-core/blob/master/patches/chrome-browser-ui-color-BUILD.gn.patch

Comment (sangwoo108) [reply patches/content-child-BUILD.gn.patch]: Sounds good! Refactored.

File: content/child/brave_runtime_features.cc [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+void BraveSetRuntimeFeaturesDefaultsAndUpdateFromArgs(
+    const base::CommandLine& command_line) {
+#if BUILDFLAG(ENABLE_PLAYLIST)
+  // When MediaSource API is enabled, we'll get "blob: " source url. We can't
+  // download media files from it, so we disable this API when playlist is
+  // eanbled.
+  blink::WebRuntimeFeatures::EnableFeatureFromString(
+      "MediaSourceStable",
+      !base::FeatureList::IsEnabled(playlist::features::kPlaylist));
+#endif
```
Comment (sangwoo108) [inline content/child/brave_runtime_features.cc:27]: @bridiver @goodov This is the new way to MediaSource API without JS injection. What do you think?

Comment (goodov) [reply content/child/brave_runtime_features.cc:27]: this approach is better, but it's still global and there's no way to turn it off other than a feature flag. It's possible that a lot of websites (conference call, streaming services, internal corporate services, etc.) may use this API to work properly. Do we want to break them?

> We're going to support a global preference or content setting for this behavior. But for now we don't have any specific plan.

Do we really need this change for all playlists to work? If it's required for few websites then why not just add it for these specific websites?

I see few options:
1. add an "allow-list" or per-website setting in this PR. 
2. postpone this exact change and make it as a standalone PR to disable this JS endpoint in the right (non-global) way.
3. hardcode (for now?) websites that need `MediaSource` hiding.

File: components/playlist/renderer/playlist_render_frame_observer.cc [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+
+PlaylistRenderFrameObserver::~PlaylistRenderFrameObserver() {}
+
+void PlaylistRenderFrameObserver::RunScriptsAtDocumentStart() {
+  if (!url_.SchemeIsHTTPOrHTTPS())
+    return;
+
+  // TODO(sko) This list should be dynamically updated from browser process.
+  // For now, we hardcode the list of domains that we want to run scripts at.
+  if (url_.host() == "www.youtube.com") {
```
Comment (goodov) [inline components/playlist/renderer/playlist_render_frame_observer.cc:27]: looks great!

but pls consider this instead of keeping `url_`:
```
static const base::NoDestructor<std::vector<blink::WebSecurityOrigin>> origins{
    {{blink::WebSecurityOrigin::Create(GURL("https://www.youtube.com"))}}};

render_frame()->GetWebFrame()->GetSecurityOrigin()->IsSameOriginWith(...);
```

Comment (sangwoo108) [reply components/playlist/renderer/playlist_render_frame_observer.cc]: Cool! Done.

File: test/base/browser_tests_main.cc [hunk 20]
```diff
@@ -31,6 +37,12 @@ int main(int argc, char** argv) {
   }
 #endif  // BUILDFLAG(IS_WIN)
 
+#if BUILDFLAG(ENABLE_PLAYLIST)
+  base::ScopedAddFeatureFlags scoped_feature_flags(
+      base::CommandLine::ForCurrentProcess());
+  scoped_feature_flags.DisableIfNotSet(playlist::features::kPlaylist);
```
Comment (goodov) [inline test/base/browser_tests_main.cc:43]: this looks weird. what are we trying to do here? feature should be disabled anyway, because it's disabled by default.

File: browser/ui/webui/playlist_page_handler.cc [hunk 21]
```diff
@@ -72,6 +76,20 @@ void PlaylistPageHandler::AddMediaFilesFromPageToPlaylist(const std::string& id,
...
+    const std::string& playlist_id) {
+  auto* browser = chrome::FindBrowserWithWebContents(web_contents_);
+  DCHECK(browser);
+  auto* tab_strip_model = browser->tab_strip_model();
+  for (auto i = 0; i < tab_strip_model->count(); i++) {
+    if (auto* contents = tab_strip_model->GetWebContentsAt(i);
+        contents != web_contents_) {
+      GetPlaylistService(profile_)->RequestDownloadMediaFilesFromContents(
+          playlist_id, contents);
```
Comment (goodov) [inline browser/ui/webui/playlist_page_handler.cc:88]: this approach to enumerate only browser-bound web_contents is not pretty, but as a wip solution for the feature development it's ok.
I don't know what the final result we want so I'm not sure what to recommend here, but I hope this will be transformed into something more reliable in future.

File: components/playlist/playlist_download_request_manager.cc [hunk 22]
```diff
@@ -159,21 +146,28 @@ void PlaylistDownloadRequestManager::DidFinishLoad(
...
-  };
+void PlaylistDownloadRequestManager::OnGetMedia(
+    base::WeakPtr<content::WebContents> contents,
+    base::Value value) {
+  if (!contents)
+    return;
+
+  DCHECK(!callback_for_current_request_.is_null()) << " callback already ran";
+  auto callback = std::move(callback_for_current_request_);
+  DCHECK(callback_for_current_request_.is_null());
```
Comment (goodov) [inline components/playlist/playlist_download_request_manager.cc:170]: I'm pretty sure `std::move()` will work, we don't need to debug flying neutrinos 😄. let's remove this line.

File: components/playlist/playlist_download_request_manager.cc [hunk 23]
```diff
@@ -159,21 +146,28 @@ void PlaylistDownloadRequestManager::DidFinishLoad(
...
-    LOG(ERROR) << value;
-  };
+void PlaylistDownloadRequestManager::OnGetMedia(
+    base::WeakPtr<content::WebContents> contents,
+    base::Value value) {
+  if (!contents)
+    return;
+
+  DCHECK(!callback_for_current_request_.is_null()) << " callback already ran";
+  auto callback = std::move(callback_for_current_request_);
```
Comment (goodov) [inline components/playlist/playlist_download_request_manager.cc:169]: the whole idea of having a single download operation and assuming that the class-bound callback will always be mapped to this download seems pretty fragile. We may want to improve this in next PRs.

File: browser/ui/webui/playlist_page_handler.cc (hunk 21 above)
Comment (sangwoo108) [reply browser/ui/webui/playlist_page_handler.cc:88]: Yeah. I agree. There's no entry point for now and this is all experimental. We'll see :)

File: components/playlist/playlist_download_request_manager.cc (hunk 22 above)
Comment (sangwoo108) [reply components/playlist/playlist_download_request_manager.cc]: Yes 😆. Removed.

File: components/playlist/playlist_download_request_manager.cc (hunk 23 above)
Comment (sangwoo108) [reply components/playlist/playlist_download_request_manager.cc:168]: I truly agree. Each request should process its own pipeline. Not like this.

**Review by goodov — APPROVED** [review]

**Comment by sangwoo108** [comment]
HI, @diracdeltas `ExecuteJavascript(ForTesting)` from browser process was replaced with RenderFrameObserver as @bridiver suggested. Do you have anything else for me to address?

**Comment by sangwoo108** [comment]
Thank you all for reviewing this :)
